### PR TITLE
Refactor OpenThread Global Variables into a Single Context Structure

### PR DESCRIPTION
**Pull request by nibanks**
In order to support multiple OpenThread instances in the same process and to support Windows Kernel Mode, all the global and static variables had to be removed. This change moves them all into the otContext structure. A pointer for the context is then passed around (and stored in some classes) to that everything has access to the necessary data, for that instance.

### CODE REVIEW

File: output/include/openthread.h [hunk 1]
```diff
@@ -0,0 +1,1701 @@
...
+ * 
+ * @param[in]    aContextBuffer      The buffer for OpenThread to use for allocating the otContext structure.
+ * @param[inout] aContextBufferSize  On input, the size of aContextBuffer. On output, if not enough space for otContext, 
+                                     the number of bytes required for otContext.
+ *
+ * @retval otContext*  The new OpenThread context structure.
+ *
+ */
+otContext *otEnable(void *aContextBuffer, uint64_t *aContextBufferSize);
+
```
Comment (nibanks) [inline output/include/openthread.h]: I think it might be good to bring back otInit to allocate, and initialize the otContext, and make otEnable/otDisable just manage the mEnabled state. What do you guys think?

File: src/core/common/message.hpp [hunk 2]
```diff
@@ -90,6 +90,7 @@ struct MessageListEntry
 struct BufferHeader
 {
     class Buffer *mNext;  ///< A pointer to the next Message buffer.
+    otContext *mContext;  ///< A pointer to the OpenThread context.
```
Comment (nibanks) [inline src/core/common/message.hpp]: Should we save the context with every buffer or should we pass the context into every Message function?

**Comment by nibanks** [comment]
I can't figure out why all the certification tests fail. Any suggestions as to how to debug these issues?

Also, the GitHub diff shows several files as just added lines where I actually just modified the file (openthread.h & openthread-types.h). I am unsure why this is.

**Comment by nibanks** [comment]
I fixed the line endings by adding a .gitattributes file to indicate that all .h, .hpp, .c, .cpp files should use LF line endings (unix style). The diffs now show up correctly.

**Comment by darconeous** [comment]
I will take a closer look at this sometime next week.

**Comment by nibanks** [comment]
I signed it!

**Comment by nibanks** [comment]
It would be great if someone could look at this sooner rather than later. I have to spend a couple of hours each time I merge the latest master because of all the conflicts and build breaks.

File: examples/platforms/posix/uart.c [hunk 3]
```diff
@@ -188,7 +188,7 @@ void posixUartUpdateFdSet(fd_set *aReadFdSet, fd_set *aWriteFdSet, int *aMaxFd)
 
 void posixUartProcess(void)
 {
-    const int flags = POLLRDNORM | POLLERR | POLLNVAL | POLLHUP;
+    const int flags = POLLIN | POLLRDNORM | POLLERR | POLLNVAL | POLLHUP;
```
Comment (darconeous) [inline examples/platforms/posix/uart.c]: A little curious about this change. It looks like an oversight on our part (Not sure why we weren't doing `POLLIN` to begin with). Were you running into problems without this change? It might be worthwhile to break this out into a separate pull request. (If you do, just remove `POLLRDNORM`, since `POLLIN` implies it)

File: include/cryptocontext.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
+/*
+ *  Copyright (c) 2016, Nest Labs, Inc.
```
Comment (darconeous) [inline include/cryptocontext.h]: Wrong copyright.

File: examples/platforms/posix/uart.c (hunk 3 above)
Comment (nibanks) [inline examples/platforms/posix/uart.c]: When I run the CLI on the Windows Subsystem for Linux, it didn't support `POLLRDNORM`, so I had to add this to get it to work.

File: include/cryptocontext.h [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    // Variables
+    //
+
+    bool mIsInitialized;
+    unsigned char mMemoryBuf[MBED_MEMORY_BUF_SIZE];
+
+    mbedtls_aes_context mAesContext;
+    mbedtls_md_context_t mSha256Context;
+
+} otCryptoContext;
```
Comment (darconeous) [inline include/cryptocontext.h]: If we are going through the effort of moving to contexts, we should have individually-typed cryptographic contexts for each type of cryptographic operation instead of a onesie-context for all operations. This makes the API much more flexible.

File: include/openthread.h [hunk 6]
```diff
@@ -157,83 +163,123 @@ const char *otGetVersionString(void);
...
+
+/**
+ * This function disables the OpenThread library.
+ *
+ * Call this function when OpenThread is no longer in use.
+ *
+ * @param[in] aContext  The OpenThread context structure.
+ *
+ */
+void otFreeContext(otContext *aContext);
```
Comment (darconeous) [inline include/openthread.h]: I like your constructor semantics. Let's rename these functions:

* `otInit()` ➔ `otContextInit()`
* `otFreeContext()` ➔ `otContextFinalize()`

File: include/openthread.h [hunk 7]
```diff
@@ -157,83 +163,123 @@ const char *otGetVersionString(void);
...
  *
  */
-ThreadError otEnable(void);
+ThreadError otEnable(otContext *aContext);
 
 /**
  * This function disables the OpenThread library.
  *
- * Call this function when OpenThread is no longer in use.  The client must call otEnable() to use OpenThread
+ * This function disables OpenThread processing. The client must call otEnable() to use OpenThread
```
Comment (darconeous) [inline include/openthread.h]: 👍

File: include/openthread.h (hunk 6 above)
Comment (jwhui) [inline include/openthread.h]: `otEnable()` and `otDisable()` were intended to serve the purpose of `otInit` and `otFreeContext` proposed here.  I see you have repurposed `otEnable()` to call `otInterfaceUp()` and `otThreadStart()`.  I'm not sure we need to provide a convenience function that collapses the two calls.

**Comment by darconeous** [comment]
By the way, in case you are following alone:

 I'll be going through and making a lot of markups as I review. It would be best to avoid acting on them until I'm finished reviewing, because after I'm done I might have advice which might make those efforts wasted. I'll let you know when I'm finished.

File: src/cli/cli.cpp [hunk 8]
```diff
@@ -1520,6 +1520,11 @@ void Interpreter::ProcessScan(int argc, char *argv[])
     AppendResult(error);
 }
 
+void Interpreter::s_HandleActiveScanResult(otActiveScanResult *aResult)
+{
+    Uart::sUartServer->mInterpreter.HandleActiveScanResult(aResult); // BUGBUG - Callbacks must provide a context
```
Comment (darconeous) [inline src/cli/cli.cpp]: Can you elaborate a little more on this comment? As I'm taking this all in it isn't immediately obvious to me what specifically the intended next action would be.

File: src/core/common/logging.hpp [hunk 9]
```diff
@@ -41,10 +41,22 @@
 #include <openthread-core-config.h>
 #include <platform/logging.h>
 
+#ifdef WINDOWS_LOGGING
```
Comment (darconeous) [inline src/core/common/logging.hpp]: These changes seem unrelated to this pull request.

File: src/cli/cli.cpp (hunk 8 above)
Comment (nibanks) [inline src/cli/cli.cpp]: All registered callbacks must have a callback context pointer. This BUGBUG was a note to myself to make another change to add this support.

File: src/core/common/logging.hpp (hunk 9 above)
Comment (nibanks) [inline src/core/common/logging.hpp]: It did get brought in when I split all my changes up. I can remove it for now.

File: include/openthread.h (hunk 6 above)
Comment (nibanks) [inline include/openthread.h]: I was also a bit unsure of how to proceed with the `mEnabled` state in general. As long as we have an `mEnabled` state, we should have a function that corresponds to setting it. If we just have the design that there is no such thing as a disabled otContext, then we can get rid of the variable, and the corresponding functions.

File: src/core/openthreadcontext.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+#include <stdint.h>
+#include <stdbool.h>
+
+#include <openthread-types.h>
+#include <thread/thread_netif.hpp>
+#include <net/ip6_mpl.hpp>
+#include <net/ip6_routes.hpp>
+
+#ifndef C_ASSERT
+#define C_ASSERT(e) typedef char __C_ASSERT__[(e)?1:-1]
```
Comment (darconeous) [inline src/core/openthreadcontext.h]: Cool! Never seen this trick before.

File: examples/apps/ncp/main.c [hunk 11]
```diff
@@ -26,31 +26,43 @@
...
-void otSignalTaskletPending(void)
+void otSignalTaskletPending(otContext *aCtx)
 {
+    (void)aCtx;
 }
 
 int main(int argc, char *argv[])
 {
+    otContext *sContext;
+    uint8_t otContextBuffer[OT_CONTEXT_SIZE];
```
Comment (darconeous) [inline examples/apps/ncp/main.c]: Wouldn't we need to make sure that the context buffer is aligned?

File: examples/apps/cli/main.c [hunk 12]
```diff
@@ -26,31 +26,43 @@
...
-void otSignalTaskletPending(void)
+void otSignalTaskletPending(otContext *aCtx)
 {
+    (void)aCtx;
 }
 
 int main(int argc, char *argv[])
 {
+    otContext *sContext;
+    uint8_t otContextBuffer[OT_CONTEXT_SIZE];
```
Comment (darconeous) [inline examples/apps/cli/main.c]: This needs to be made static.

File: examples/apps/ncp/main.c (hunk 11 above)
Comment (darconeous) [inline examples/apps/ncp/main.c]: Also, this needs to be made static so that it isn't allocated from the stack.

**Comment by darconeous** [comment]
First, I want to say thanks for contributing this! This obviously took a considerable amount of work and effort. I can tell that you put a lot of thought into these changes. 

I've reviewed all of the code and my comments are outlined above. Most (all?) of the issues are minor.

I was originally concerned what the impact on code size would be with a change on this scale, so I built the CC2538 target both with and without your changes, and then ran `size` for the `ot-ncp` example:

  `TEXT`| `DATA` | `BSS` | Total  | Total (Hex)    | Commit
--------|--------|-------|--------|----------------|-----------
 100736 |   2168 | 19488 | 122392 |        0x1de18 | 139e85f (Without Contexts)
 101016 |   2160 |  3148 | 106324 |        0x19f54 | e8eacd8 (With Contexts)

The `BSS` section has decreased in size dramatically, which is likely due to you putting `otContext` on the stack(See associated comment in source code). More importantly, it looks like there is not a significant increase in `TEXT` size (just 0.28%) due to this change, which is fantastic. 

It is my opinion that the changes reflected here are ultimately where OpenThread needs to be. The OpenThread API has been a bit short-sighted in this respect and I think these changes represent a marked improvement.

That being said, the forklift-nature of this API change presents a lot of practical challenges for us which I'm not immediately sure how best to resolve. The single biggest issue is at hand is that this change totally breaks all of our internal code which uses OpenThread. In other words, if this pull request is merged all of our stuff breaks. We would need to prepare things on our side first, and that takes engineering resources away from working on our internal milestones.

We need to discuss this more internally. More information to follow.

**Comment by nibanks** [comment]
Thanks for the code review. I will address the comments you made and await your decision from your internal discussions.

File: include/cryptocontext.h (hunk 5 above)
Comment (nibanks) [inline include/cryptocontext.h]: Right now, they share the same mMemoryBuf. If I put them into different contexts, I would have to duplicate that buffer or complicate the interfaces by passing the buffers in. I also don't know this code much so I am unsure of the best way to split these.

**Comment by nibanks** [comment]
Again, Travis is having issues. Here is my branch's [build](https://travis-ci.org/nibanks/openthread/builds/153567884) with the changes.

**Comment by darconeous** [comment]
What is #425? Why not just force push to this branch, or add a merge?

**Comment by nibanks** [comment]
I wanted to separate out the change so it wouldn't be blocked waiting for this. Once it is merged into master, I will update this change to get the latest master and fix the merge conflicts.

**Comment by darconeous** [comment]
Sorry, I got a bit cross-eyed and it looked like the same changes at first glance. No worries!

**Comment by nibanks** [comment]
Merged the latest master; including the new callback context changes.

**Comment by darconeous** [comment]
Just wanted to give you a quick update. We are actively discussing this internally. The general consensus seems to be converging toward trying to figure out how to break up this task into more manageable chunks that can be more easily integrated with minimal disruption. I feel like we are likely going to converge on an action plan over the weekend. I'll post an update with more details on Monday.

Thanks for your patience!

**Comment by nibanks** [comment]
Any updates? I'm not sure there is any easy way to make this change smaller. If you have existing code using this, the quickest way to updated it would just create a new static buffer & otContext for the time being and use that everywhere.

**Comment by darconeous** [comment]
Hey Nick, sorry for the late response.

We are currently discussing an entirely different approach to solving this problem that would require almost no API changes and would add no additional overhead in the embedded case.

The idea is to make the API context stateful by introducing the following new API calls:

```
struct otInstancePrivate;
typedef struct otInstancePrivate otInstance;

/// Initializes the given instance.
/** Returns the required size of the instance. The instance
 *  is only initialized if the return value is greater than or equal
 *  to aInstanceSize. */
ssize_t otInstanceInit(otInstance *aInstance, ssize_t aInstanceSize);

/// Finalize the given instance.
/** Called when the instance is no longer needed. */
void otInstanceFinalize(otInstance *aInstance);

/// Sets the current instance for this thread.
/** Must be called before any API calls are made on a given thread.
 *  On embedded systems that only ever have one instance, this
 *  call is a no-op. */
void otInstanceSet(otInstance *aInstance);

/// Gets the current instance for this thread.
/** Returns the instance previously set on the current thread with
 *  otInstanceSet(). Returns NULL if otInstanceSet() was never called
 *  on the given thread. On embedded systems, this function always
 *  returns the global instance. */
otInstance* otInstanceGet(void);
```

This approach requires thread-local storage if you want your platform to support multiple independent OpenThread instances. What's nice is that for platforms where support for multiple independent instances is not required, the additional overhead is minimal (and can even be made zero with some preprocessor macros).

So, the question then becomes: does the windows kernel allow for thread-local storage in device drivers?

**Comment by nibanks** [comment]
As far as I know Windows kernel mode doesn't have any built in thread local variable support. Ignoring Windows for a second, how would you define the thread local instance exactly? Would you do some thing like this?

`thread_local osInstance sCurrentInstance;`

I am almost positive that won't work. Everything in Windows passes context pointer around, and most globals/statics are discouraged if not out right disallowed.

**Comment by nibanks** [comment]
I just tried adding a `thread_local` variable and I get the following error from VS:

`C2949        thread_local is not supported with /kernel`

**Comment by nibanks** [comment]
The `thread_local `version of the API could be provided on top of the interface I have implemented. And only those who want it would compile it into their code. But it would require duplicating a lot of stuff.

**Comment by darconeous** [comment]
We were thinking more about using an API that provides that support rather than a language-defined capability. For example, on platforms that support pthreads we would use the `pthread_key_create()`, `pthread_getspecific()`, and `pthread_setspecific()` APIs. It looks like Windows user-mode has thread-local storage support.

After doing a bit of Googling, it looks like this design pattern might not work well in the Windows kernel.

**Comment by darconeous** [comment]
Given that the above plan isn't going to work for the Windows kernel, let's scrap it and go with plan B.

The plan B involves three pull requests:

1. New pull request with just the API changes, with the API implementations largely ignoring the instance pointer. Everything would still use globals internally. This allows us to review and approve *just* the API changes independently of other changes. It also allows us to go ahead and get started on the work to switch over to the new API without having to worry about destabilizing our own code. Also, let's use `otInstance` instead of `otContext`, since we generally use the term 'context' to refer to a void pointer we can get passed to callbacks.
2. There are some APIs that don't need access to the primary `otInstance` structure since they are actually methods to independent objects. The message pool and crypto APIs come to mind. The changes necessary to support this would happen at this step. This pull request would include API changes and internal changes.
3. Implement internal support for `otInstance`. This pull request would include no API changes.

This approach allows for the library to be usable at each step and amortizes the integration workload over a period of time. This makes life much easier on our end.

Thoughts?

**Comment by nibanks** [comment]
A few questions:

For number 1 above, since the signature of the API is still changing, all the consumers of the API in openthread will have to be updated. What do you want to do with them? Add the code to call into `otInstanceInit` and pass the resulting otInstance in, or do you just want to pass in NULL for that change? 

For number 2, I am a bit unclear what is happening. Since some of the functions don't need the `otInstance`, wouldn't they just not be updated in number 1?

**Comment by darconeous** [comment]
> For number 1 above, since the signature of the API is still changing, all the consumers of the API in openthread will have to be updated. What do you want to do with them? Add the code to call into `otInstanceInit` and pass the resulting otInstance in, or do you just want to pass in NULL for that change? 

The former: The consumers would be updated to act like they are doing something when under the hood they aren't actually doing anything.

> For number 2, I am a bit unclear what is happening. Since some of the functions don't need the `otInstance`, wouldn't they just not be updated in number 1?

They wouldn't need the `otInstance` specifically, but they are still operating on some sort of object that is conceptually independent from `otInstance`. This commit is where those objects are fleshed out.

**Comment by nibanks** [comment]
> They wouldn't need the otInstance specifically, but they are still operating on some sort of object that is conceptually independent from otInstance. This commit is where those objects are fleshed out. 

What I have seen is that everything comes back to `otInstance`. The way they reference it comes down to one of two solutions:

1. The `otInstance` is passed into their functions.
2. The `otInstance` is passed into their constructor and they store a pointer to it internally.

Currently, for `Message` (and a number of internal classes) I was using #2 above. This is why the OpenThread APIs related to `Message` don't have the pointer passed in. I feel like you have to make that decision in your 'step 1'. I guess I still don't see a clear distinction between 1 & 2.

**Comment by darconeous** [comment]
The message pool API mention came from some offline thoughts from @jwhui. I'll let him comment on that.

With regard to the crypto API, leaving that for step 2 gives me some time to rethink the matter. More on this later.

Ultimately, step 1 is about getting the API hammered out, with perhaps a few choice exceptions. Code flow should largely remain unchanged. It may be the case that it makes more sense to go straight to step 3. I think that depends on @jwhui's thoughts.

**Comment by nibanks** [comment]
Any updates on if you want to split the change or not? I sent a PR with just the API changes but haven't gotten any feedback yet.

**Comment by darconeous** [comment]
I've been out sick since yesterday, I haven't looked over your latest PR. I'll poke others to have a look.

File: include/openthread-types.h [hunk 13]
```diff
@@ -44,6 +44,18 @@
...
+#pragma GCC diagnostic ignored "-Winvalid-offsetof"
+#define CONTAINING_RECORD(address, type, field) \
+    ((type *)((uint8_t*)(address) - offsetof(type, field)))
+#pragma GCC diagnostic pop*/
+#define BASE 0x1
+#define myoffsetof(s,m) (((size_t)&(((s*)BASE)->m))-BASE)
+#define CONTAINING_RECORD(address, type, field) \
+    ((type *)((uint8_t*)(address) - myoffsetof(type, field)))
+#endif
+
```
Comment (nibanks) [inline include/openthread-types.h]: I had problems with warnings on Linux builds. Any idea on how to get the above to work, without hacking the `BASE` variable in?

File: src/core/mac/mac.cpp [hunk 14]
```diff
@@ -48,13 +48,14 @@
 #include <platform/random.h>
 #include <thread/mle_router.hpp>
 #include <thread/thread_netif.hpp>
+#include <openthreadinstance.h>
+
```
Comment (nibanks) [inline src/core/mac/mac.cpp]: Remove extra line

File: src/core/openthreadinstance.h [hunk 15]
```diff
@@ -0,0 +1,92 @@
+/*
+ *  Copyright (c) 2016, Microsoft Corporation.
```
Comment (nibanks) [inline src/core/openthreadinstance.h]: Update copyright

File: include/openthread-types.h (hunk 13 above)
Comment (darconeous) [inline include/openthread-types.h]: Hmm... What are the warnings?  Adding the `BASE` macro really fixes it? Not sure what you were trying to work around with `myoffsetof()`...

Comment (darconeous) [inline include/openthread-types.h]: Still curious what the warnings were...?

Comment (nibanks) [inline include/openthread-types.h]: `error: invalid access to non-static data member ‘otInstance::mIp6’  of NULL object [-Werror=invalid-offsetof]`
It doesn't like accessing a `NULL` pointer?

File: doc/spinel-protocol-src/Makefile [hunk 16]
```diff
@@ -1,92 +0,0 @@
-#
```
Comment (nibanks) [inline doc/spinel-protocol-src/Makefile]: I'm not sure why this file was removed. I will fix it.

File: tests/unit/test_diag.cpp [hunk 17]
```diff
@@ -67,7 +67,6 @@ extern "C" void otPlatRadioReceiveDone(otInstance *, RadioPacket *aFrame, Thread
     (void)aError;
 }
 
-
```
Comment (nibanks) [inline tests/unit/test_diag.cpp]: Remove changes to this file.

File: third_party/mbedtls/repo/programs/pkey/rsa_priv.txt [hunk 18]
```diff
@@ -1,8 +1,8 @@
-N = A1D46FBA2318F8DCEF16C280948B1CF27966B9B47225ED2989F8D74B45BD36049C0AAB5AD0FF003553BA843C8E12782FC5873BB89A3DC84B883D25666CD22BF3ACD5B675969F8BEBFBCAC93FDD927C7442B178B10D1DFF9398E52316AAE0AF74E594650BDC3C670241D418684593CDA1A7B9DC4F20D2FDC6F66344074003E211
```
Comment (nibanks) [inline third_party/mbedtls/repo/programs/pkey/rsa_priv.txt]: Revert changes to this file.

File: third_party/mbedtls/repo/programs/pkey/rsa_pub.txt [hunk 19]
```diff
@@ -1,2 +1,2 @@
-N = A1D46FBA2318F8DCEF16C280948B1CF27966B9B47225ED2989F8D74B45BD36049C0AAB5AD0FF003553BA843C8E12782FC5873BB89A3DC84B883D25666CD22BF3ACD5B675969F8BEBFBCAC93FDD927C7442B178B10D1DFF9398E52316AAE0AF74E594650BDC3C670241D418684593CDA1A7B9DC4F20D2FDC6F66344074003E211
```
Comment (nibanks) [inline third_party/mbedtls/repo/programs/pkey/rsa_pub.txt]: Revert changes to this file.

File: src/core/openthreadinstance.h [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    Thread::Crypto::MbedTls mMbedTls;
+    Thread::Ip6::Ip6 mIp6;
+    Thread::ThreadNetif mThreadNetif;
+
+    // Constructor
+    otInstance(void);
+
+} otInstance;
+
+__inline otInstance *otInstanceFromIp6(Thread::Ip6::Ip6 *aIp6)
```
Comment (darconeous) [inline src/core/openthreadinstance.h]: `__inline` is technically vendor specific. I recommend just using `static inline`, since this is C++ and `inline` is guaranteed to be a viable keyword.

**Comment by darconeous** [comment]
@nibanks: Let us know if you are satisfied with the current state of the diff. If so, I'll toss this over to @jwhui to merge.

**Comment by nibanks** [comment]
I am satisfied with this. I have my branches derived from these changes and they are working great in my Windows driver.

**Comment by jwhui** [comment]
This looks great.  One minor nit: how about using `openthread-instance.h` rather than `openthreadinstance.h`?
